Netwayste Version 2

Manghi opened the pull request:
Changes for the Netwayste Re-architecture as related to #151 and #153 .

Goal is to make the Conwayste networking code easier to maintain and grow by reducing the coupling between modules, features, and implementation.

AaronM04 commented:
I feel really good about this branch! :D

Mind if I push changes or would you prefer that I just review as you work on it?

Manghi commented:
> I feel really good about this branch! :D
> 
> Mind if I push changes or would you prefer that I just review as you work on it?

@AaronM04 
Commit and push as you'd like!

AaronM04 commented on netwaystev2/src/transport/transport.rs:
```diff
@@ -0,0 +1,231 @@
+        ))
+    }
+
+    pub async fn monitor(&mut self) -> Result<()> {
```
Can we name this something else, like `run`, since it's doing more than just monitoring. I would expect something that "monitors" to be a read-only thing...

AaronM04 commented on netwaystev2/src/transport/transport.rs:
```diff
@@ -21,14 +21,14 @@ use tokio::time as TokioTime;
-type TransportRspRecv = Receiver<TransportRsp>;
+pub type TransportRspRecv = Receiver<TransportRsp>;
 
 type TransportNotifySend = Sender<TransportNotice>;
```
Should these be `pub` as well?

Manghi replied:
Possibly. Right now it's only those that cross the Filter/Transport bridge.

AaronM04 commented on netwaystev2/src/filter/sortedbuffer.rs:
```diff
@@ -0,0 +1,251 @@
+        for tuple in &self.heap {
+            if tuple.0.0 == sequence {
+                count += 1;
+                sequence += 1;
```
Will this wrap correctly without panicking? @Manghi

AaronM04 replied:
Also, I noticed the tests are failing for this because there is no way to iterate over the heap in the order that `pop()` will happen (without consuming the heap, which would be inefficient). I wonder if we really need this operation since we may not be able to implement it in an efficient way?

Manghi replied:
> without consuming the heap, which would be inefficient
Exactly!

Thinking some more about this, you're absolutely right, we don't *really* need the operation. There aren't any great sorted set libs presently available for our use case anyways.

I think comparing against the `sequence` via `peek()` before a `pop()` in a loop is the right way to go for now. This is corrected in the latest commit.

AaronM04 commented on netwaystev2/src/filter/sortedbuffer.rs:
```diff
@@ -0,0 +1,251 @@
+}
+
+#[cfg(test)]
+mod test {
```
What do you think about moving all these tests to a separate file, and in general, keeping our tests separate from our business logic?

Manghi replied:
Love the idea. We tend to write quite a bit of tests anyways which makes the files unmanageable during navigation.

Manghi replied:
Needed to make one function public in order to access it. I tried `pub(crate)` first but that didn't work since tests are seen as external to the crate.

AaronM04 replied:
Well... they don't have to be _that_ separate... you have them as integration tests: https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests

That said, it's not a big deal that `contains_sequence_number` is public now.

Manghi replied:
I was thinking about this this morning and decided to go with `src/filter/tests` instead. Will be in latest set of commits once I push.

AaronM04 commented on netwaystev2/src/filter/interface.rs:
```diff
@@ -0,0 +1,242 @@
+    pub in_progress:  bool,
+}
+
+pub enum FilterEndpointData {
```
What do you think about putting this somewhere else? I feel like this and the error below it don't belong in `interface.rs` since they are internal to the Filter.

Manghi replied:
Yep agreed! There's a bunch of stuff here that doesn't belong -- rather, a lot of it is tailed towards the app-layer. I'll move it out.

Manghi replied:
I created a `src/protocol` folder which contains the data structure definitions for `Packet` and the like. Let me know what you think!

Manghi replied:
Wasn't sure if I should move `PingPong` too but I decided to leave it in `filter.rs` for now.

AaronM04 replied:
I really like the organization in `src/protocol`. I can actually find structs and enums with this.

The only thing that I would change is this:

```rust
        /* PR_GATE add in later
        game_updates:    Vec<GameUpdate>, // Information pertaining to a game tick update.
        universe_update: UniUpdate,       // TODO: add support
        */
```
Can you make the complete structure we have in the `netwayste` crate be here too (no temporarily commented data structures) please?

Also I am not sure why we need a `Default` implementation for `Packet...?

AaronM04 replied:
The other thought I had is maybe we could merge `response.rs` into `request.rs`. Otherwise there will need to be an `updatereply.rs` file added to be consistent, and that seems like a lot of files. 😁

Manghi replied:
Removed the `PR_GATE` comments and added in their associated data structures. 

I like that `request.rs` and `response.rs` are separate. The idea was that any data structures used within a `Packet` variant's would be defined in their respective files for quick look up. `UpdateReply` uses only primitive data types so there wasn't a need to define an `updatereply.rs`.... yet.

AaronM04 commented:
It turns out we forgot a `Shutdown` command to shut all the layers down! I'm going to add this to the docs now. The idea is it propagates from the top down (app to filter to transport), causing everything to exit ASAP. We might want the server app layer in particular to have another `GentleShutdown` which notifies players and gives them time to finish their game while not allowing any new connections.

I also need to start adding some of the interface between filter and app layers.

Manghi commented:
The test case `basic_server_filter_flow()` does not expire on my system. It hangs indefinitely. Does it work for you?

```bash
....
conwayste/netwaystev2 $ cargo test 
test filter::tests::sortedbuffer_tests::seqminheap_insert_sequential_with_gaps ... ok
test filter::tests::time_advance_tests::time_advancing_works ... ok
test filter::tests::time_advance_tests::basic_server_filter_flow has been running for over 60 seconds
^C
conwayste/netwaystev2 $ 
```

AaronM04 commented:
> The test case `basic_server_filter_flow()` does not expire on my system. It hangs indefinitely. Does it work for you?
> 
> ```shell
> ....
> conwayste/netwaystev2 $ cargo test 
> test filter::tests::sortedbuffer_tests::seqminheap_insert_sequential_with_gaps ... ok
> test filter::tests::time_advance_tests::time_advancing_works ... ok
> test filter::tests::time_advance_tests::basic_server_filter_flow has been running for over 60 seconds
> ^C
> conwayste/netwaystev2 $ 
> ```

I ran it several times and sometimes it fails while other times it hangs forever. I was running the wrong test (`cargo test -p netwaystev2 time_advance_test`) and got excited that it was passing and not hanging.  I'll take a look tonight.

AaronM04 commented on netwaystev2/src/filter/filter.rs:
```diff
@@ -0,0 +1,756 @@
+        tokio::pin!(filter_notice_tx);
+
+        let mut ping_interval_stream = tokio::time::interval(Duration::new(2, 0));
+        ping_interval_stream.tick().await;
```
@Manghi adding this line causes the `basic_server_filter_flow` test to fail.

Manghi replied:
Fixed.

AaronM04 replied:
Nice catch with the `tokio::task::yield_now`!

AaronM04 reviewed (commented):
plz fix

Manghi commented on netwaystev2/src/filter/filter.rs:
```diff
@@ -422,6 +425,17 @@ impl Filter {
+                        }
+                        ResponseCode::LeaveRoom => {
+                            server.game_update_seq = None;
+                        }
```
You should be able to condense it via an  `|` operator.

Manghi commented on netwaystev2/src/filter/filter.rs:
```diff
@@ -451,17 +465,87 @@ impl Filter {
+                }
+
+                let mut start_idx = None;
+                match (server.game_update_seq, game_update_seq) {
```
Can you leave a comment explaining what the purpose of this block is? This seems like one of those things that makes sense now but will be easy to forget after several months.

Manghi commented on netwaystev2/src/filter/filter.rs:
```diff
@@ -451,17 +465,87 @@ impl Filter {
+                                error!("[FILTER] failed to process game update {:?}: {}", game_updates[i], e);
+                            }
+                        }
+                        server.game_update_seq.as_mut().map(|seq| *seq += 1); // Increment
```
Comment is a bit useless. Perhaps explain why the increment is needed?

Manghi commented on netwaystev2/src/filter/client_update.rs:
```diff
@@ -1,29 +1,136 @@
 
 pub struct ClientRoom {
     player_name:       String, // Duplicate of player_name from OtherEndServer (parent) struct
+    player_id:         Option<usize>, // If player is not a lurker, must be Some(...) before the first GenStateDiff
```
Would you be opposed to aliasing `Option<usize>` to `PlayerID`?

Manghi commented on netwaystev2/src/filter/client_update.rs:
```diff
@@ -1,29 +1,136 @@
+                }
+                let uni = big_bang.birth()?;
+                let game = ClientGame {
+                    player_id:     self.player_id, // copy
```
What's the comment for?

AaronM04 replied:
It's copying the `player_id` from the room to the game.

Manghi commented on netwaystev2/src/filter/client_update.rs:
```diff
@@ -1,29 +1,136 @@
+                self.player_id = None;
+                let mut new_other_players = HashMap::new();
+                for name in self.other_players.keys() {
+                    new_other_players.insert(name.clone(), None);
```
Why does everyone become a lurker on `GameFinish`?

AaronM04 replied:
Just an arbitrary choice. We can revisit when exploring the actual game play. :)

Manghi commented on netwaystev2/src/filter/client_update.rs:
```diff
@@ -1,29 +1,136 @@
+                for player in players {
+                    if player.name == self.player_name {
+                        // Hey, that's us!
+                        self.change_own_player_id(player.index);
```
Why does the index change here on a `PlayerList` command? Is that if the server assigned the player a different Universe index?

AaronM04 replied:
Yes, that's the reason.

AaronM04 commented on netwaystev2/src/transport/endpoint.rs:
```diff
@@ -0,0 +1,272 @@
+        for (endpoint, container) in &mut self.transmit {
+            for PacketContainer { packet, info, tid } in container {
+                // Add the packet to the list of retriable packets if enough time has passed since the last transmission
+                if info.transmit_interval != Duration::ZERO
```
Well then it would keep sending it as fast as possible. :smile:

migdall commented on nwv2-python-wrapper/src/protocol/update.rs:
```diff
@@ -0,0 +1,54 @@
+use pyo3::exceptions::*;
```
Is that a Rust import?

AaronM04 replied:
It is! It's importing from this: https://docs.rs/pyo3/latest/pyo3/exceptions/index.html

These are Rust wrappers for some standard Python exception classes like `Exception`, `ValueError`, etc.

Also hi 👋

AaronM04 committed:
Filter: send FilterRsp in response to FilterCmd; also unify error types

Manghi committed:
Status packet wrapper simplified; added concepts for status getters

AaronM04 committed:
Merge branch 'master' into mang/netwayste_v2

AaronM04 committed:
minor Filter notebook changes

AaronM04 committed:
FilterNoticeW/ResponseCodeW: rework some getters

AaronM04 committed:
PacketW status returns appropriate type for each member (not String)

AaronM04 committed:
Filter layer: keep track of cookies for client-side operations, once logged in

Manghi committed:
PacketW GetStatus returns appropriate type too

Manghi committed:
section headers in BasicPacket, and expanded TODO

Manghi committed:
add getter for packet request fields

Manghi committed:
wip: assert request fields

Manghi committed:
Fixed BasicPacket assertion tests for request action comparison

AaronM04 committed:
filter: minor comment

AaronM04 committed:
Finish Packet wrapper (add GenPartInfoW, UniUpdateW, respesponse, update, updatereply)

AaronM04 committed:
Filter: remove unused fields from OtherEndServer

AaronM04 committed:
save Jupyter notebook for Filter

AaronM04 committed:
Transport: new notice -- EndpointIdle

AaronM04 committed:
Transport: show EndpointIdle in Jupyter notebook

AaronM04 committed:
Troubleshooting Transport layer bug added in commit 4347023

AaronM04 committed:
Changes to my debugging notebook

AaronM04 committed:
Filter: refactoring

AaronM04 committed:
More troubleshooting Transport layer bug added in commit 4347023

AaronM04 committed:
Fix the Transport layer bug added in commit 4347023 :)

The problem was that for the SendPackets Transport command, errors were
being swallowed and _no_ TransportRsp was being sent in reply. This
broke everything.

AaronM04 committed:
cargo fmt

AaronM04 committed:
Filter bug fix: only send NewChats notice up to App layer if there are actually new chats

AaronM04 committed:
Filter bug fix: default retry interval was way too low (retrying way too often)

AaronM04 committed:
deleting my debugging notebook now that debugging is complete

AaronM04 committed:
Refactor Transport layer: process_transport_command SendPackets handling

This should make it harder to accidentally create a bug like I did
before.

AaronM04 committed:
Saving WrappedFilter notebook

AaronM04 committed:
WIP Client and Server notebook for Filter layer

Manghi committed:
filter must create a new transport endpoint on connect

Manghi committed:
wip: (hangs) notebook update for client and server filters

AaronM04 committed:
Filter: if server and receive new Connect packet, send Transport a NewEndpoint cmd

AaronM04 committed:
Filter: client mode: on EndpointIdle notice, send KeepAlive to server

Manghi committed:
(hangs): server filter notification handling

AaronM04 committed:
Python wrapping: enable netwaystev2 Rust logging in Python, fix hang

Manghi committed:
LTI colors client and server log events

Manghi committed:
added succinct logging metatags in filter and transport layers

Manghi committed:
resolved some old and new build warnings

Manghi committed:
ran cargo fmt

Manghi committed:
remove dev comment

AaronM04 committed:
Transport: minor bug fix and change to new_endpoint

AaronM04 committed:
Add local_addr to Transport layer

AaronM04 committed:
Changes to reduce excessive numbers of packets:

* Increase the endpoint timeout the Filter layer gives the Transport layer from 250ms to 5s

* Increase ResponseCode packet retry intervals to 9999 seconds (effectively
  preventing retries) as a hack since the PacketSettings struct doesn't have a
  field for turning off or limiting retries.

AaronM04 committed:
Update notebook output

It's still panicking because of the FilterInterface logic @manghi added:

  panicked at 'failed to acquire notify rx lock: TryLockError(())', nwv2-python-wrapper/src/filter/mod.rs:221:18

Manghi committed:
separated log event type from log event origin; updated missing instances

Manghi committed:
Added comment describing the log message format in nwv2

AaronM04 committed:
un-doc the comment to fix compile error

AaronM04 committed:
Filter: minor refactor

AaronM04 committed:
Transport: the NewEndpoint cmd now upserts and does not ever return error

Manghi committed:
commented out handle_filter_notification; added split command/response

Manghi committed:
additional getters on wrappers; bug fix to allow keywords

Manghi committed:
ran cargo fmt

Manghi committed:
notebook successfully working for client connect and server response

The client port was hardcoded so that an endpoint wrapper could be
created for it and used in the response by the 'server'. This will need
to be swapped out for pulling the port from the bounded socket at
runtime.

The notebook will still sometimes fail on the assertion that the notify
lists are non-empty when getting them, but rerunning these cells acts as
a workaround. Adding in a `sleep(n)` call does not resolve the issue.

Manghi committed:
removed dead code per working changes made today

AaronM04 committed:
Minor fixes to FilterInterface

AaronM04 committed:
[WIP] Fix the Filter command_response hang.....

AaronM04 committed:
Merge branch 'master' into mang/netwayste_v2

Manghi committed:
filter run() terminates on errors; removed bad usage of filter-rsp

Manghi committed:
clientandserver-filter: use command_response on fifaces

Manghi committed:
fix convience wrapper default argument in clientandserver-filter

Manghi committed:
wip: clientandfilter-notebook, new and join room updates

AaronM04 committed:
update older Jupyter notebooks with latest code to ensure nothing broke

AaronM04 committed:
Filter layer: all FilterCmd handling results in FilterRsp now

AaronM04 committed:
WIP Filter layer ping stuff; also minor tweaks to Transport layer

AaronM04 committed:
[WIP] Filter ping-related changes:

* Increase ping frequency by 10x from 2s to 200ms
* Do not send KeepAlive to server on idle if this is a ping endpoint
* Allow servers to process pings
* Logging changes

Open questions:

* Why is there no PingResult?
* Why does logging all seem to happen after sending ClearPingEndpoints?

AaronM04 committed:
Filter: ping bug fix; reduced latency filter ping depth

AaronM04 committed:
Minor cleanup

AaronM04 committed:
Python wrapper: shorten timestamp in log lines

AaronM04 committed:
Filter pings: add noisy log message to fix tonight

AaronM04 committed:
save work for Filter pings notebook

AaronM04 committed:
Added some random Py async code to mysteriously unblock the Filter-Pings notebook

AaronM04 committed:
Transport: do not put non-retriable packets in transmit queue (no need to DropPacket)

AaronM04 committed:
add a comment

AaronM04 committed:
Filter: do not send DropPacket for non-retriable packets; minor refactoring

AaronM04 committed:
Filter: allow multiple in-flight pings

AaronM04 committed:
Filter: simplify pings and remove "active ping in progress"

Manghi committed:
wip: working through issues with lfi/lti missing

AaronM04 committed:
[WIP] [DNC] More progress on removing LTI from FilterInterface

AaronM04 committed:
Removed LTI ("Python sandwich") from FilterInterface

AaronM04 committed:
Filter: ChangeServerStatus command; still need to send it and add to nwv2 wrapper

AaronM04 committed:
nwv2 wrapper: remove commented leftovers from Python sandwich

AaronM04 committed:
Support ChangeServerStatus filter cmd in nwv2 wrapper and use in Jupyter notebook

Manghi committed:
removed WrappedFilter: no more sandwiched TransportInterafce

Manghi committed:
Verified BasicPacket runs and passes; cleared output

Manghi committed:
Verified ClientAndServer-Filter tests pass; cleared output

Manghi committed:
Verified Transport tests pass; cleared output

Manghi committed:
Verified Filter-Pings tests pass; cleared output

Manghi committed:
Verified Filter tests pass; cleared output

Manghi committed:
removed remaining interface from ClientAndServer-Filter; passes

Manghi committed:
[filter] wrapped logging macros to prefix execution context

Manghi committed:
[transport] added TransportMode and python wrapper

so that we can use it to print context in log statements

Manghi committed:
[transport] wrapped logging macros to prefix execution context

Manghi committed:
conditionally compile context logging prefixes in nwv2

Useful to enable prefixes only within the nwv2-python-wrapper test
cases. This will avoid adding extra unnecessary 'c' and 's' statements
when the client and server are executed by a user.

AaronM04 committed:
Rustfmt; no other changes

AaronM04 committed:
Reduce warnings in netwaystev2 and nwv2-python-wrapper crates

Manghi committed:
Addressed warnings related to the use of a deprecated constant in nwv2

Warning resolved is:

```
warning: use of deprecated constant `pyo3::impl_::deprecations::PYMETHODS_ARGS_ATTRIBUTE`: the `#[args]` attribute for `#[methods]` is being replaced by `#[pyo3(signature)]`
   --> nwv2-python-wrapper/src/protocol/packet.rs:134:7
    |
134 |     #[args(member = "\"pong_nonce\"")]
    |       ^^^^
    |
    = note: `#[warn(deprecated)]` on by default

warning: use of deprecated constant `pyo3::impl_::deprecations::PYMETHODS_ARGS_ATTRIBUTE`: the `#[args]` attribute for `#[methods]` is being replaced by `#[pyo3(signature)]`
   --> nwv2-python-wrapper/src/protocol/packet.rs:157:7
    |
157 |     #[args(member = "\"sequence\"")]
    |       ^^^^

```

AaronM04 committed:
Minor cleanup in Filter layer (most important first)

AaronM04 committed:
Correct a FilterCmd variant type

AaronM04 committed:
Correct a FilterCmd variant type -- also fix the wrapper

AaronM04 committed:
Comments

AaronM04 committed:
Comment

AaronM04 committed:
Revising Filter notebook to test login flow [WIP]; may have bug with excessive EndpointIdle from Transport layer

Manghi committed:
implemented FilterCmd::DropEndpoint

Manghi committed:
Merge branch 'master' into mang/netwayste_v2

AaronM04 committed:
Fix Transport layer bug where EndpointIdle notice sent too often

AaronM04 committed:
Filter notebook: auth flow and keepalive handling on server-side

Manghi committed:
foundation for netwayste v2 server binaries

Manghi committed:
pretty-print nw-server config

AaronM04 committed:
Delete useless appendage!

Manghi committed:
added control section for specifying unix socket path

AaronM04 committed:
Filter: fix auth (almost) and numerous tidying up and fixes; also tests pass [WIP]

AaronM04 committed:
Filter: lots of refactoring and rearranging; also worked on CompleteAuthRequest [WIP]

AaronM04 committed:
Filter: micro-optimization lolz

AaronM04 committed:
Filter: server validates cookie received from client

AaronM04 committed:
Filter: very minor cleanup -- should not be behavior changes

AaronM04 committed:
Filter: very minor cleanup -- should not be behavior changes (similar change to previous one)

AaronM04 committed:
Filter: server-side: store last sent ResponseCode and re-send on new Packet::Request if needed [WIP]

AaronM04 committed:
Filter: incomplete and untested resend_and_drop_enqueued_response_codes [WIP]

AaronM04 committed:
Filter: resend_and_drop_enqueued_response_codes; needs unit testing

AaronM04 committed:
Filter: bug fix to resend_and_drop_enqueued_response_codes

AaronM04 committed:
Filter: another bug fix to resend_and_drop_enqueued_response_codes

AaronM04 committed:
Filter: bug fix to send_response_code related to above; sigh....

AaronM04 committed:
netwaysted & netwaystectl - Unix control socket

AaronM04 committed:
netwaysted & netwaystectl - cleanup and improvements

AaronM04 committed:
netwaysted & netwaystectl - cargo +nightly fmt

Manghi committed:
configure tracing as the loggnig crate

Manghi committed:
'contract' module ensures server + client agree on messaging protocol

Manghi committed:
working example of server and client messaging

Manghi committed:
server: clean up socket gracefully on SIGTERM and SIGINT

Manghi committed:
formatting and long comment split

AaronM04 committed:
netwaysted & netwaystectl - ability to define subcommands in netwaystectl

AaronM04 committed:
Filter: quick todo

AaronM04 committed:
Filter: complete logic around sending FilterNotice::NewResponseCode

AaronM04 committed:
Filter: very minor cleanup -- no behavior changes

AaronM04 committed:
Filter: Send FilterNotice::EndpointTimeout on TransportNotice::EndpointTimeout

AaronM04 committed:
Filter: bug & performance fixes for sending NewGenStateDiff

AaronM04 committed:
Filter: wrap up FilterNotice::NewGameUpdates and a bug fix to save last full generation

AaronM04 committed:
Filter: start on UpdateReply; bug fix to GameUpdate::RoomDeleted handling

AaronM04 committed:
delete empty server.rs

AaronM04 committed:
Filter (server-side): lay groundwork for handling Update/UpdateReply

AaronM04 committed:
Filter (server-side): ClientRoom create/destroy

One more ClientRoom destroy case to handle (RoomDeleted).

Manghi committed:
deleted unused 'enum-tree' subproject

Manghi committed:
wip: started dissector-netwaystev2 for packet capturing nwv2 traffic

Manghi committed:
deserialize up until the good stuff

Manghi committed:
migrated dissect-netwaystev2 into the conwayste org, conwayste-tools

Manghi committed:
and remove the workspace entry too

AaronM04 committed:
Lower MTU by 32 bytes since it's just a guess

AaronM04 committed:
Filter (server-side): saving my progress on Update/UpdateReply [WIP]

AaronM04 committed:
Filter (server-side): minor improvement to error handling in prev. commit

AaronM04 committed:
Filter (server-side): GameUpdateQueue

Manghi committed:
server: add a module to manage a pool of rooms for alloc/free

AaronM04 committed:
Filter: saving my changes [WIP]

AaronM04 committed:
Filter: cargo +nightly fmt

Manghi committed:
use a trace event for every config dump line

Manghi committed:
Merge branch 'master' into mang/netwayste_v2

Manghi committed:
libraries do not specify the frontend logger

Manghi committed:
server: wip, add netwayste layers (registration failing)

AaronM04 committed:
cargo +nightly fmt

AaronM04 committed:
Allow private servers

AaronM04 committed:
server: fix bug with registering server

AaronM04 committed:
server: use [D] prefix for logging

Manghi committed:
Revert "server: use [D] prefix for logging"

This reverts commit 4c047988fadae5f0008be93a67cae488a2410c11.

[D] was meant to indicate daemon logging, but this is a bit redundant
because tracing emits a statement which includes the binary name.

Manghi committed:
improve comment docs on registrar url and print config

Manghi committed:
enable compatibility between tracing and log; minimize external crate log statements

Manghi committed:
cargo +nightly fmt

Manghi committed:
grammar improvement in comment

AaronM04 committed:
app layer: a closed channel will always trigger a select! -- 100% CPU :(

AaronM04 committed:
server: remove some unnecessary clones() and use Server mode for Filter layer

Manghi committed:
nw-server: removed unigen code from app-layer. revisit concept

Manghi committed:
refactor into new control message handling

Manghi committed:
nw-server: cargo add bincode

Manghi committed:
nw-server: implement structured messaging between the control and daemon

Manghi committed:
add a ping endpoint for the registrar when in server mode

Manghi committed:
do not restrict pings to just the client

Manghi committed:
commented manual ip since host lookup is failing

Manghi committed:
Revert "commented manual ip since host lookup is failing"

This reverts commit f99c3cf2a7b674aea387fd23247d6ce92dc90eb1.

Manghi committed:
Revert "do not restrict pings to just the client"

This reverts commit 2621f1a14bde8d3ac58153392e015d31e119cf5f.

Manghi committed:
Revert "add a ping endpoint for the registrar when in server mode"

This reverts commit e96ef29d060f0f7162ba9d293605d665deb4c348.

AaronM04 committed:
Merge remote-tracking branch 'origin/master' into mang/netwayste_v2

Manghi committed:
nw-server: propagate the server name from config

AaronM04 committed:
cargo +nightly fmt

AaronM04 committed:
Filter (server-side): game updates (untested; still need to handle acks)

AaronM04 committed:
Filter (client-side): accommodate the change in meaning of None `game_update_seq` in the protocol

AaronM04 committed:
Transport: cleanup: remove old definition of UDP_MTU_SIZE (now in common.rs)

AaronM04 committed:
Filter (server-side): game updates -- bug fixes

AaronM04 committed:
Filter (server-side): bug fixes

AaronM04 committed:
Filter (server-side): game updates (untested)

AaronM04 committed:
nwv2 wrapper: archived notebook broken due to RequestAction changes (ClientAuthRequest)

AaronM04 committed:
Filter: KeepAlives no longer involve App layer

AaronM04 committed:
Filter: fix nasty bug in previous commit -- unfortunately the fix is complicated

AaronM04 committed:
Transport: minor cleanup

AaronM04 committed:
Filter: tell Transport layer to drop endpoint when unexpected packet received

AaronM04 committed:
Filter: move some client state from Filter to OtherEndServer

AaronM04 committed:
Filter notebook: test a GameUpdate (currently not being acked right by client)

AaronM04 committed:
Filter(client): bug fix to GameUpdate handling

AaronM04 committed:
nwv2-python-wrapper readme improvements

AaronM04 committed:
Filter: working on server uni updates [WIP]

AaronM04 committed:
Filter: server uni updates complete but untested; turns out I need to finish client-side uni update

AaronM04 committed:
Filter: finish client uni update: partial_gen; untested

AaronM04 committed:
comment

AaronM04 committed:
libconway: generate diffs for use in Jupyter notebook

AaronM04 committed:
ggez client: remove use of `pub` in GameArea

AaronM04 committed:
Filter: protocol change! GenStateDiffPart is now Vec<u8> not string, and LZ4-compressed [DNC]

netwaystev2 crate compiles; nwv2-python-wrapper crate does not yet

AaronM04 committed:
Python wrapper: fix GenStateDiffPartW after protocol change

AaronM04 committed:
New Jupyter notebook for uni updates [WIP]

AaronM04 committed:
UniUpdates notebook: combine and simplify

Manghi committed:
relocated room.rs from server daemon to server app layer

Manghi committed:
removed pool concept from rooms; fixed broken unit test

Manghi committed:
add getters for a room

Manghi committed:
wip: add room to server app layer

Manghi committed:
skeleton for request action processing in server app layer

Manghi committed:
ran fmt

Manghi committed:
implemented new room request action handling in server app layer

Manghi committed:
significantly simplified server room design

Manghi committed:
renamed RoomList -> RoomStatus

Manghi committed:
no longer panic when response codes are unhandled in server app

AaronM04 committed:
libconway: make all BigBang fields public

AaronM04 committed:
ggez client: fix a weird little thing

AaronM04 committed:
netwaystev2: cargo +nightly fmt

AaronM04 committed:
libconway: finish diff example

AaronM04 committed:
UniUpdates notebook: processes a diff

AaronM04 committed:
UniUpdates notebook: minor change to remove a placeholder

AaronM04 committed:
Filter (client-side): properly ignore outdated GenStateDiffParts from server

Manghi committed:
wip: adding support for daemon to query the app layer for status and control

Manghi committed:
add tabled crate for pretty printing

Manghi committed:
fill in first app response message

Manghi committed:
support table view of room statuses in ctl and daemon

Because the pretty print consumes a lot more characters in the daemon
response message, I increased the size of the socket to 4KB.

Manghi committed:
cargo +night fmt

Manghi committed:
scaffolding for converting control commands into AppCmd

AaronM04 committed:
cargo +nightly fmt

AaronM04 committed:
Filter(server): make GameUpdateQueue generic --> UnackedQueue<GameUpdate>; also for chats now

AaronM04 committed:
protocol: comments

AaronM04 committed:
filter: start on server-side chat support; also minor cleanup

AaronM04 committed:
filter (server-side): bug fix to an unlikely error message

AaronM04 committed:
filter (server-side): starting on chat support [WIP]

Manghi committed:
rename room module to indicate it stores several of them

Manghi committed:
cleaning up some hastily written code for clarity

AaronM04 committed:
filter (client-side): bug fix -- drop room if we successfully left room

AaronM04 committed:
filter (server-side): chat support - UNTESTED

AaronM04 committed:
filter (server-side): major improvement and cleanup for GenStateDiff sending

Manghi committed:
Authenticate player connections in the server app layer

AaronM04 committed:
nwv2 wrapper changes now that AuthDecision does not have cookie

AaronM04 committed:
comment

AaronM04 committed:
Chats notebook

Manghi committed:
address build warning about resolver inconsistencies

Manghi committed:
wip: add noteboko specifically for connect authentication

AaronM04 committed:
update readme now that I tested it on Debian

AaronM04 committed:
Remove netwayste v1 from ggez client

AaronM04 committed:
Very beginning of Client App layer [WIP]
